### PR TITLE
Add the ST_ConvexHull function

### PR DESCRIPTION
**Pull request by tpeng-fb**
Add `ST_ConvexHull(geometry)` function to return the minimum convex geometry that encloses all points in the input geometry.

Due to a bug in `OGCGeometry#convexHull` API, https://github.com/Esri/geometry-api-java/issues/172, current implementation doesn't support geometry collections.

### CODE REVIEW

**Comment by electrum** [comment]
Please update the commit message to match the standard style: https://chris.beams.io/posts/git-commit/

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 1]
```diff
@@ -133,6 +133,9 @@ Accessors
 
     Returns the point value that is the mathematical centroid of a geometry.
 
+.. function:: ST_ConvexHull(Geometry) -> Geometry
+    Returns the convex hull of a geometry represents the minimum convex geometry that encloses all geometries within the set.
```
Comment (electrum) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: This should have a newline before the description (see others for examples).

The sentence is not valid. I think you meant to say “a geometry that represents”

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 2]
```diff
@@ -126,6 +126,23 @@ public void testSTCentroid()
         assertFunction("ST_AsText(ST_Centroid(ST_GeometryFromText('POLYGON ((0 0, 0 5, 5 5, 5 0, 0 0), (1 1, 1 2, 2 2, 2 1, 1 1))')))", VARCHAR, "POINT (2.5416666666666665 2.5416666666666665)");
     }
 
+    @Test
+    public void testSTConvexNull()
+    {
+        // test Invalid types
```
Comment (electrum) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: “invalid” should not be uppercase

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 3]
```diff
@@ -198,6 +199,16 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(createFromEsriGeometry(centroid, geometry.getEsriSpatialReference()));
     }
 
+    @Description("Returns the minimum convex geometry that encloses all geometries within the set")
+    @ScalarFunction("ST_ConvexNull")
```
Comment (electrum) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: “null” should be “hull”

File: presto-docs/src/main/sphinx/functions/geospatial.rst (hunk 1 above)
Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: For clarity, let's say "Returns the minimum convex geometry that encloses all input geometries."

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 4]
```diff
@@ -198,6 +199,16 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(createFromEsriGeometry(centroid, geometry.getEsriSpatialReference()));
     }
 
+    @Description("Returns the minimum convex geometry that encloses all geometries within the set")
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: "all input geometries" might be more clear

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 5]
```diff
@@ -198,6 +199,16 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @ScalarFunction("ST_ConvexNull")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice stConvexNull(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: s/Null/Hull here and a few places throughout this PR

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 6]
```diff
@@ -198,6 +199,16 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    {
+        OGCGeometry geometry = deserialize(input);
+        validateType("ST_IsConvexNull", geometry, EnumSet.of(MULTI_LINE_STRING, MULTI_POINT, MULTI_POLYGON, GEOMETRY_COLLECTION));
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: The SQL/MM Part 3 standard defines ST_ConvexHull function for all types of geometries, not only multi types and geometry collections. Hence, this check is not needed.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 7]
```diff
@@ -126,6 +126,23 @@ public void testSTCentroid()
...
+        assertInvalidFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('LINESTRING EMPTY')))", VARCHAR, "ST_Polygon only applies to MULTI_LINE_STRING or MULTI_POINT or MULTI_POLYGON or GEOMETRY_COLLECTION\n. Input type is: LINE_STRING");
+        assertInvalidFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('POINT (3 5)')))", VARCHAR, "ST_Polygon only applies to MULTI_LINE_STRING or MULTI_POINT or MULTI_POLYGON or GEOMETRY_COLLECTION\n. Input type is: POINT");
+        assertInvalidFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('LINESTRING (1 1, 2 2, 3 3)')))", VARCHAR, "ST_Polygon only applies to MULTI_LINE_STRING or MULTI_POINT or MULTI_POLYGON or GEOMETRY_COLLECTION\n. Input type is: LINESTRING");
+        assertInvalidFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('POLYGON ((1 1, 1 4, 4 4, 4 1)')))", VARCHAR, "ST_Polygon only applies to MULTI_LINE_STRING or MULTI_POINT or MULTI_POLYGON or GEOMETRY_COLLECTION\n. Input type is: POLYGON");
+
+        // test valid types
+        assertFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('MULTILINESTRING EMPTY')))", VARCHAR, "POLYGON EMPTY");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Could you add tests for all types of empty geometries? E.g. POINT EMPTY, MULTIPOINT EMPTY, LINESTRING EMPTY, MULTILINESTRING EMPTY, POLYGON EMPTY, MULTIPOLYGON EMPTY, GEOMETRYCOLLECTION EMPTY.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 8]
```diff
@@ -126,6 +126,23 @@ public void testSTCentroid()
...
+        assertFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('MULTILINESTRING (1 1, 5 1), (2 4, 4 0) (2 -4, 4 4) (3 -2, 4 -3)')))", VARCHAR, "POLYGON (1 1, 2 4, 4, 4, 5 1, 4 -3, 2 -4)");
+        assertFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('MULTIPOINT (0 2, 1 0, 3 0, 4 0, 4 2, 2 2, 2 4)')))", VARCHAR, "POLYGON (0 2, 2 4, 4 2, 4 0, 1 0)");
+        assertFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('MULTIPOLYGON (0 3, 2 0, 3 6), (2 1, 2 3, 5 3, 5 1), (1 7, 2 4, 4 2, 5 6, 3 8)')))", VARCHAR, "POLYGON (0 3, 1 7, 3 8, 5 6, 5 1, 2 0)");
+        assertFunction("ST_AsText(ST_ConvexNull(ST_GeometryFromText('GEOMETRYCOLLECTION (POINT (2 3), LINESTRING (2 8, 7 10)), POINT (8 10), POLYGON (4 4, 4 8, 9 8, 6 6, 6 4, 8 3, 6 1), POINT (4 2), LINESTRING (3 6, 5 5), POLYGON (7 5, 7 6, 8 6, 8 5)')))", VARCHAR, "POLYGON (2 3, 2 8, 7 10, 8 10, 9 8, 8 3, 6 1)");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Could you add tests for single geometries, both convex and non-convex?

Comment (tpeng-fb) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: @mbasmanova Hi, I think I understand the concept of convex hull for polygon. But what's the convex hull for a point and a line string? I don't know if it's applicable.

**Comment by tpeng-fb** [comment]
@mbasmanova  Hi I made some changes according to the suggestion but I am not sure if it's valid or not. Could you check it a little bit?

**Comment by tpeng-fb** [comment]
@mbasmanova  I made those test passed.
Seems like convexHull would have 3 characters:
1. it would not accept empty geometry
2. it would return a invalid ring that formats a single line if the points / strings/ polygon was in the same line.
2. it would return mutipolygon empty if all points / strings polygon was in a single point.

**Comment by mbasmanova** [comment]
@tpeng-fb For the case when all points are on the same line, can we get the following behavior described in https://en.wikipedia.org/wiki/Convex_hull?

"If the points of S are all on a line, the convex hull is the line segment joining the outermost two points."

E.g. `convex-hull(POINT(1 2)) -> POINT(1,2)` and `convex-hull(MULTIPOINT ((1 1), (2 2))) -> LINESTRING ((1 1), (2 2))`

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 9]
```diff
@@ -198,6 +200,64 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+        OGCGeometry geometry = deserialize(input);
+        if (!isValidGeometryForConvexHull(geometry)) {
+            return serialize(geometry);
+        }
+        OGCGeometry convexHull = geometry.convexHull();
+        long pointCount = getOGCGeometryPointCount(convexHull);
+        if (pointCount == 0) {
+            return serialize(getSinglePointFromGeometry(geometry));
+        }
+        else if (pointCount == 2) {
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: This looks like a bug in the ESRI's library. I submitted an issue: https://github.com/Esri/geometry-api-java/issues/172

Thanks for coming up with a workaround. 

tranformOGCGeometryToOGCLineString function is very short and is used only once. Let's inline it here.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 10]
```diff
@@ -198,6 +200,64 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice stConvexHull(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (!isValidGeometryForConvexHull(geometry)) {
+            return serialize(geometry);
+        }
+        OGCGeometry convexHull = geometry.convexHull();
+        long pointCount = getOGCGeometryPointCount(convexHull);
+        if (pointCount == 0) {
```
[the remark this answers is not in the document]
Comment (tpeng-fb) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova  I think this would happen when everything was in a single point.
You can see all those examples in the test.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 11]
```diff
@@ -198,6 +200,64 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
         return serialize(createFromEsriGeometry(centroid, geometry.getEsriSpatialReference()));
     }
 
+    @Description("Returns the minimum convex geometry that encloses all input geometries")
+    @ScalarFunction("ST_ConvexHull")
+    @SqlType(GEOMETRY_TYPE_NAME)
+    public static Slice stConvexHull(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (!isValidGeometryForConvexHull(geometry)) {
```
[the remark this answers is not in the document]
Comment (tpeng-fb) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova I think in theory you are right. But `convexHull` method would not accept any EMPTY. Do I need to filter it out and rebuild the input?

[the remark this answers is not in the document]
Comment (tpeng-fb) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova I think in theory you are right. But convexHull method would not accept any EMPTY. Do I need to filter it out and rebuild the input?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @tpeng-fb Got it. Generally speaking all the geospatial functions assume that input geometries are simple and valid (see the first paragraph in the docs below). Hence, we don't need to worry about input geometries that are *all single point* except when these are *exactly one point*.  E.g. we don't need to have `convexHull` function work on geometries like `MULTILINESTRING ((1 1), (1 1), (1 1, 1 1))` and `MULTIPOINT (1 1, 1 1, 1 1)`.

https://prestodb.io/docs/current/functions/geospatial.html

[the remark this answers is not in the document]
Comment (tpeng-fb) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova  I see, I can just revise it back

[the remark this answers is not in the document]
Comment (tpeng-fb) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova Just want to double check if we need to care about something like MUTIPOINT (1 1) or POLYGON (1 1)

**Comment by tpeng-fb** [comment]
@mbasmanova Hi Masha, thanks for all the suggestions, I made the change for the last two suggestions on my local. But I want to amend my commit after I have figured out the questions I had replied to you.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Having an empty geometry as part of a geometry collection doesn't make that collection invalid or not simple. ST_ConvexHull function needs to process such collections without an error. 

Adding the logic to check if geometry collection has any empty elements and, if so, rebuild it to exclude these before calling `convexHull` method feels like an overkill. 

It looks like ESRI folks have a patch that addresses all the issues you found - Esri/geometry-api-java#172 . 

One option would be to wait for the patch to get released, upgrade the dependency and implement ST_ConvexHull using updated `convexHull` API. Another option is to add a simple check for geometry type and throw invalid-argument exception for all geometry collections and document this limitation in this PR, then upgrade ESRI and remove the workarounds and the limitation in a future PR.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: `MUTIPOINT (1 1)` and `GEOMETRY_COLLECTION (POINT (1 1))` are valid geometries. `POLYGON (1 1)` is not.

[the remark this answers is not in the document]
Comment (tpeng-fb) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova  In this situation I think this private method is still valid. But I think we can skip the checking for COLLECTION in case we take the option two you have mentioned above.

**Comment by tpeng-fb** [comment]
@mbasmanova I have revised this pr again to follow the second option.
I did not remove getSinglePointFromGeometry method since we need to generate POINT (1 1) for MULTPOINT (1 1)

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 12]
```diff
@@ -198,6 +198,29 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+    {
+        OGCGeometry geometry = deserialize(input);
+        validateType("ST_ConvexHull", geometry, EnumSet.of(POINT, MULTI_POINT, LINE_STRING, MULTI_LINE_STRING, POLYGON, MULTI_POLYGON));
+        if (geometry.isEmpty()) {
+            return input;
+        }
+        OGCGeometry convexHull = geometry.convexHull();
+        long pointCount = getOGCGeometryPointCount(convexHull);
+        if (pointCount == 0) {
+            return serialize(getSinglePointFromGeometry(geometry));
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: It would be more efficient to handle points explicitly and using `isEmpty()` check in place of `pointCount == 0`. I'm explaining why in a comment attached to `getOGCGeometryPointCount`.

It would also be helpful to explain why these workarounds are needed and provide pointers to ESRI issues.

```
if (geometry.isEmpty()) {
   return input;
}

if (GeometryType.getForEsriGeometryType(geometry.geometryType()) == POINT)  {
   return input;
}

OGCGeometry convexHull = geometry.convexHull();
if (convexHull.isEmpty()) {
   // This happens for a single-point multi-point because of a bug in ESRI library - https://github.com/Esri/geometry-api-java/issues/172
   return serialize(createFromEsriGeometry(((MultiVertexGeometry) geometry.getEsriGeometry()).getPoint(0), null));
}

if (convexHull instanceOf MultiVertexGeometry) {
    MultiVertexGeometry multiVertex = (MultiVertexGeometry) convexHull;
    if (multiVertex. getPointCount() == 2) {
       // This happens when all points of the input geometry are on the same line because of a bug in ESRI library - https://github.com/Esri/geometry-api-java/issues/172
        ...
    }
}

return serialize(convexHull);
```

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 13]
```diff
@@ -133,6 +133,10 @@ Accessors
 
     Returns the point value that is the mathematical centroid of a geometry.
 
+.. function:: ST_ConvexHull(Geometry) -> Geometry
+
+    Returns the minimum convex geometry that encloses all input geometries.
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: Could you mention that geometry collections are not supported yet?

```
Returns the minimum convex geometry that encloses all input geometries.
This function doesn't support geometry collections.
```

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 14]
```diff
@@ -308,7 +331,7 @@ public static Double lineLocatePoint(@SqlType(GEOMETRY_TYPE_NAME) Slice lineSlic
         }
 
         GeometryType pointType = GeometryType.getForJtsGeometryType(point.getGeometryType());
-        if (pointType != GeometryType.POINT) {
+        if (pointType != POINT) {
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: This is unrelated change. Could you move it into a separate commit?

**Comment by mbasmanova** [comment]
Hi Tianshu,

Would you update PR description to

* remove "Added the first commit ..." and "Instructions" sections
* simplify the "Limitations" section to list specific limitations of which I believe there is only one: current implementation doesn't support geometry collections?

Perhaps,

```
Add `ST_ConvexHull(geometry)` function to return the minimum convex geometry that encloses all points in the input geometry.

Due to a bug in `OGCGeometry#convexHull` API, https://github.com/Esri/geometry-api-java/issues/172, current implementation doesn't support geometry collections.
```

**Comment by tpeng-fb** [comment]
@mbasmanova 
Hi Masha, I revised the pr according to your suggestions. Please check again.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 15]
```diff
@@ -623,7 +677,17 @@ public void testSTGeometryType()
         assertFunction("ST_GeometryType(ST_GeometryFromText('MULTIPOINT (1 1, 2 2)'))", VARCHAR, "ST_MultiPoint");
         assertFunction("ST_GeometryType(ST_GeometryFromText('MULTILINESTRING ((1 1, 2 2), (3 3, 4 4))'))", VARCHAR, "ST_MultiLineString");
         assertFunction("ST_GeometryType(ST_GeometryFromText('MULTIPOLYGON (((1 1, 1 4, 4 4, 4 1)), ((1 1, 1 4, 4 4, 4 1)))'))", VARCHAR, "ST_MultiPolygon");
-        assertFunction("ST_GeometryType(ST_GeometryFromText('GEOMETRYCOLLECTION(POINT(4 6),LINESTRING(4 6,7 10))'))", VARCHAR, "ST_GeomCollection");
+        assertFunction("ST_GeometryType(ST_GeometryFromText('GEOMETRYCOLLECTION(POINT(4 6),LINESTRING(4 6, 7 10))'))", VARCHAR, "ST_GeomCollection");
         assertFunction("ST_GeometryType(ST_Envelope(ST_GeometryFromText('LINESTRING (1 1, 2 2)')))", VARCHAR, "ST_Polygon");
     }
+
+    private void assertConvexHull(String inputWKT, String expectWKT)
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: nit: would you move this function and `assertConvexHullInvalidFunction` right after `testSTConvexHull`?

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 16]
```diff
@@ -198,6 +198,35 @@ public static Slice stCentroid(@SqlType(GEOMETRY_TYPE_NAME) Slice input)
...
+        OGCGeometry convexHull = geometry.convexHull();
+        if (convexHull.isEmpty()) {
+            // This happens for a single-point multi-point because of a bug in ESRI library - https://github.com/Esri/geometry-api-java/issues/172
+            return serialize(createFromEsriGeometry(((MultiVertexGeometry) geometry.getEsriGeometry()).getPoint(0), null));
+        }
+        if (GeometryType.getForEsriGeometryType(convexHull.geometryType()) == MULTI_POLYGON) {
+            MultiVertexGeometry multiVertex = (MultiVertexGeometry) convexHull.getEsriGeometry();
+            if (multiVertex.getPointCount() == 2) {
+                // This happens when all points of the input geometry are on the same line because of a bug in ESRI library - https://github.com/Esri/geometry-api-java/issues/172
+                OGCGeometry transformedConvexHull = createFromEsriGeometry(new Polyline(multiVertex.getPoint(0), multiVertex.getPoint(1)), null);
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: nit: transformedConvexHull -> linestring, for clarity

**Comment by tpeng-fb** [comment]
@mbasmanova 
Thanks again. I had addressed the comments.
@electrum 
Could you check again to see if it's looking good now?
